### PR TITLE
allow setting the system time via strptime

### PR DESCRIPTION
**Pull request by badbadc0ffee**
## Description:

The idea came up in PR #1441 to allow the system time to be manipulated programmatically if no timesource like sntp or homeassistant is available.
An example for the new functions to parse a time string (``ESPTime::strptime()``) and to set the system clock accordingly (``RealTimeComponent::set()``) has been added to ``tests/test1.yaml`` but the documentation is missing.

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs/pull/921)

### CODE REVIEW

**Comment by foxsam21** [comment]
Looks good. However what about a scenario where one does not have any (currently) compatible time platform, i.e. homeassistant, gps, (rtc?), etc.. and wants to still use time. This use case should work, there will be time drift, in a general sense using esphome's time component. In such a case the time would _need_ to be set manually but as it stands now it can only set if there is already one of the time platforms. 
I would suggest a stand-alone "manual" component that does not need another time platform.

**Comment by badbadc0ffee** [comment]
> I would suggest a stand-alone "manual" component that does not need another time platform.

You mean, a component that has the interface of a real time clock bit is none? I don't quite get the idea what that would be useful for. Without an external time source, the device would of course loose ist time information with every reboot.

What I'd suggest instead is to create a network time source (i.e. homeassistant and / or sntp) which will only sync the time if a connection is available. When no network connection can be established, you can set the time and date via the interface I suggested.

**Comment by foxsam21** [comment]
I'm working on a device with a clock display and I want to be able to
travel with it (after? corona) and not rely on HA or sntp for time, but
still be able to use functions of esphome and HA (alarm clock and other
status indications) when it is at home without reconfiguring anything.

I use my own manual time `custom_component` for this purpose but it's not
really good enough for a PR. I guess you have a point to just put in sntp
or whatever and it can work when it is available and be "manual" when not.

**Comment by badbadc0ffee** [comment]
> I'm working on a device with a clock display and I want to be able to travel with it [...]
So I guess you will be using some kind of RTC such as DS1307 (#1441)? 
Than you can use that component and would not need a "manual" clock component.
Otherwise the device will loose its time information on every reboot.

File: esphome/components/time/real_time_clock.h [hunk 1]
```diff
@@ -53,6 +53,19 @@ struct ESPTime {
    */
   std::string strftime(const std::string &format);
 
+  /** Update an ESPTime struct based on a string as specified by the format argument.
+   * @see https://www.gnu.org/software/libc/manual/html_node/Formatting-Calendar-Time.html#index-strftime
+   */
+  ESPTime &strptime(const char *time, const char *format) __attribute__((format(strftime, 3, 0)));
```
Comment (OttoWinter) [inline esphome/components/time/real_time_clock.h:56]: This should instead be a `static` function like `static ESPTime strptime` (then it also matches the python datetime class, which also has it as a static method: https://docs.python.org/3/library/datetime.html#datetime.datetime.strptime)

File: esphome/components/time/real_time_clock.h [hunk 2]
```diff
@@ -53,6 +53,19 @@ struct ESPTime {
...
+
+  /** Update an ESPTime struct based on a string as specified by the format argument.
+   * @see https://www.gnu.org/software/libc/manual/html_node/Formatting-Calendar-Time.html#index-strftime
+   *
+   * @warning This method uses dynamically allocated strings which can cause heap fragmentation with some
```
Comment (OttoWinter) [inline esphome/components/time/real_time_clock.h:61]: Unless you're calling this very often, it won't be an issue. ESPHome has plenty of objects being modified on the heap. We should drop the `char *` method IMO

**Pull request closed without merging**